tribits/core/package_arch/TribitsProjectWriteConfigFile.cmake

```diff
@@ -0,0 +1,228 @@
+#
+# 3. Neither the name of the Corporation nor the names of the
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+# ************************************************************************
+# @HEADER
+
+
+################################################################################
+#
+# Module TribitsInternalPackageWriteConfigFile.cmake
+#
+# This module contains code for generating <Project>Config.cmake files for
+# TriBITS projects.
+#
+################################################################################
+
+
+include(TribitsInternalPackageWriteConfigFile)
+
+
+# Write the outer TriBITS <Project>Config.cmake file
+#
+# If ${PROJECT_NAME}_VERSION is not set or is '' on input, then it will be set
+# to 0.0.0 in order to create the ${PROJECT_NAME}ConfigVersion.cmake file.
+#
+# ToDo: Finish documentation!
+#
+function(tribits_write_project_client_export_files)
+
+  set(EXPORT_FILE_VAR_PREFIX ${PROJECT_NAME})
+
+  # Reversing the package list so that libraries will be produced in order of
+  # most dependent to least dependent.
+  set(PACKAGE_LIST ${${PROJECT_NAME}_DEFINED_INTERNAL_PACKAGES})
+  if (PACKAGE_LIST)
+    list(REVERSE PACKAGE_LIST)
+  endif()
+
+  # Loop over all packages to determine which were enabled. Then build a list
+  # of all their libraries/includes in the proper order for linking
+  set(FULL_PACKAGE_SET "")
+  set(FULL_LIBRARY_SET "")
+  foreach(TRIBITS_PACKAGE ${PACKAGE_LIST})
+    if(${PROJECT_NAME}_ENABLE_${TRIBITS_PACKAGE})
+      list(APPEND FULL_PACKAGE_SET ${TRIBITS_PACKAGE})
+      list(APPEND FULL_LIBRARY_SET ${${TRIBITS_PACKAGE}_LIBRARIES})
+    endif()
+  endforeach()
+
+  set(${PROJECT_NAME}_CONFIG_LIBRARIES ${FULL_LIBRARY_SET})
+
+  # Reversing the tpl list so that the list of tpls will be produced in
+  # order of most dependent to least dependent.
+  if (${PROJECT_NAME}_DEFINED_TPLS)
+    set(TPL_LIST ${${PROJECT_NAME}_DEFINED_TPLS})
+    list(REVERSE TPL_LIST)
+  endif()
+
+  # Loop over all TPLs to determine which were enabled. Then build a list
+  # of all their libraries/includes in the proper order for linking
+  set(FULL_TPL_SET "")
```

Comment:
I don't think that is the case.  I think calling `set(<varName>)` just removes the variable `<varName>` from the current scope which is not the same as setting it to empty.  See:

* https://cmake.org/cmake/help/v3.26/command/set.html

which says:

> Zero arguments will cause normal variables to be unset. See the [unset()](https://cmake.org/cmake/help/v3.26/command/unset.html#command:unset) command to unset variables explicitly.

Setting `set(<varName> "")` creates an empty list.  For example, running:

```
set(someList "")
list(APPEND someList a)
list(APPEND someList b)
list(APPEND someList c)
message("someList = '${someList}'")
```

prints out:

```
someList = 'a;b;c'
```

That is the way it seems to work anyway.  Let me know if you observe something different.

Reply:
> I think calling set(<varName>) just removes the variable <varName> from the current scope which is not the same as setting it to empty.

Yes, that is the case. `list(APPEND)` does support appending to a list that doesn't exist yet - it goes ahead and creates the list. I believed that starting with `set(someList "")` and then appending would result in `;a;b;c`, however, that appears to not be the case - both ways appear to work:

```
$ cat test.cmake
set(list)
list(APPEND list a)
list(APPEND list b)
list(APPEND list c)
message(STATUS "${list}")

set(list "")
list(APPEND list a)
list(APPEND list b)
list(APPEND list c)
message(STATUS "${list}")
$ cmake -P test.cmake
-- a;b;c
-- a;b;c
```

Reply:
@KyleFromKitware, it is not okay to call `set(<varName>)` and then `list(APPEND <varName> ...)`.

Consider the following code:

```
set(someList "d;e" CACHE STRING "")
message("Initail value of cache var someList: '${someList}'")

set(someList "false value")
message("Initial value of local var someList: '${someList}'")

set(someList)
#set(someList "")
message("Value of \${someList} after calling set(someList): '${someList}'")

list(APPEND someList a)
list(APPEND someList b)
list(APPEND someList c)

message("Value of someList after calling list(APPEND someList ...) several times: '${someList}'")

set(someList)
message("Value of \${someList} after calling set(someList): '${someList}'")
```

When I run that with CMake 3.20.0, I get:

```
$ cmake -P create_a_list.cmake
Initail value of cache var someList: 'd;e'
Initial value of local var someList: 'false value'
Value of ${someList} after calling set(someList): 'd;e'
Value of someList after calling list(APPEND someList ...) several times: 'd;e;a;b;c'
Value of ${someList} after calling set(someList): 'd;e'
```

The problem is that the call `set(someList)` simply removes the local variable `someList` that has the value `false value` which then exposes the cache variable `someList` with the value `d;e`.  It seems what `list(APPEND someList a)` then picks up the cache variable value `d;e` and uses it to create a local variable with that value and then appends `a` to it.

So if you change `set(someList)` to `set(someList "")` in the code above, you get the desired behavior:

```
$ cmake -P create_a_list.cmake
Initail value of cache var someList: 'd;e'
Initial value of local var someList: 'false value'
Value of ${someList} after calling set(someList ""): ''
Value of someList after calling list(APPEND someList ...) several times: 'a;b;c'
Value of ${someList} after calling set(someList ""): ''
```

So you can never safely call `set(<varName>)` and expect that to start you out with a uninitialized variable.  You always have to call `set(<varName> "")` to make sure you are creating an uninitialized/empty local variable that is safe to amend.

This is another example of why CMake needs a new language.  I don't know why this is not being more seriously discussed.

Reply:
@KyleFromKitware, NOTE, because of the issue with local vars hiding cache vars, there has to be a way to initialize an empty local list var.   But note if you change `list(APPEND someList <ele>)` to `set(someList "${someList};<ele>")` you get `';a;b;c'`.  But if you change that to `set(someList ${someList} <ele>)` you get `'a;b;c'`.  Therefore, it seems that:

```
list(APPEND someList <ele>)
```

is equivalent to:

```
set(someList ${someList} <ele>)
```

(except perhaps if there are spaces in `${someList}`?)

So with `set(someList ${someList} <ele>)` that explains the behavior with local and cache vars shown [above](https://github.com/TriBITSPub/TriBITS/pull/560#discussion_r1146398285).  When `${someList}` is evaluated, it picks up the cache var value `d;e` and the statement creates the local var `someList` with the value `d;e;a`.

CMake badly needs a linter that can flag issues like this (or a more intelligent debug runtime mode for CMake that can catch cases like this at runtime).

Or just a new language ...

Reply:
FYI: This discussion was continued in:

* https://gitlab.kitware.com/cmake/cmake/-/issues/24633